### PR TITLE
Pie title

### PR DESCRIPTION
**Pull request by codrut3**
Hi,

I tried to solve #2915 in this PR. I added an attribute `title` to the pie chart, with the option to show it either inside the hole, or outside.

Showing the title inside is easy. On the other hand, I couldn't come up with an algorithm for placing the title outside that works well in all cases. The problem is that the title may intersect labels / other pies / other images or even go out the chart. So what I do, I consider 2 cases:

- there are several pies per column. Then I scale the pie until enough space for the title is left. There is no other solution, because in the following example the pies leave no free space.
newplot

- if there is one pie per column, I place the title high enough to not intersect pulled slices, and leave space for at most one label below it. It may still intersect slice labels that were moved around, or even be just badly placed. Feel free to suggest another approach :) Maybe one could use an attribute `titleheight` and let the user decide the optimal placement?

I made some mocks, so you can see how it looks.

### CODE REVIEW

File: src/traces/pie/attributes.js [hunk 1]
```diff
@@ -180,6 +180,33 @@ module.exports = {
...
+        editType: 'calc',
+        description: [
+            'Sets the title of the pie chart.',
+            'If it is empty, no title is displayed.'
+        ].join(' ')
+    },
+    titleposition: {
+        valType: 'enumerated',
+        values: ['inhole', 'outside'],
+        dflt: 'outside',
```
Comment (alexcjohnson) [inline src/traces/pie/attributes.js]: Can we make this look a bit more like [`textposition`](https://github.com/plotly/plotly.js/blob/master/src/traces/scatter/attributes.js#L512)? I'm not sure *quite* all of those options make sense, but most of them do. What you have now corresponds to `'top center'` and `'middle center'`, and it's fine to start with just those two options, but I can certainly see wanting `'top left'`, `'top right'`, and `'bottom (left|center|right)'` in the future.

Also I'd remove the `dflt` here and make it dependent on `hole`: if it's zero, default to `'top center'`, otherwise `'middle center'`.

File: src/traces/pie/defaults.js [hunk 2]
```diff
@@ -65,6 +65,15 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         }
     }
 
+    var title = coerce('title');
+    if(title) {
+        var titlePosition = coerce('titleposition');
+
+        if(titlePosition === 'inhole' || titlePosition === 'outside') {
```
Comment (alexcjohnson) [inline src/traces/pie/defaults.js]: You don't need this `if` - after `coerce`, we know `titlePosition` has a valid value. (but if you want to prohibit center positioning when there's no hole, this would be the place:
`if(!hole && titlePosition === 'middle center') traceOut.titlePosition = 'top center';`

File: src/traces/pie/plot.js [hunk 3]
```diff
@@ -454,6 +501,72 @@ function transformOutsideText(textBB, pt) {
...
+    return {
+        x: cd0.cx,
+        y: cd0.cy - (1 + maxPull) * cd0.r - (titleSpace / 2),
+        scale: Math.min(scaleX, scaleY)
+    };
+}
+
+function isSinglePie(trace) {
+    // check if there is a single pie per y-column
+    if(trace.domain.y[0] === 0 && trace.domain.y[1] === 1) return true;
```
Comment (alexcjohnson) [inline src/traces/pie/plot.js]: I don't think we want to treat single-pie-per-column differently from multiple, but it's not clear to me offhand whether to shrink all or none. The argument for shrinking all of them is that there's also `layout.title` at the top, and we don't want pie titles to overlap with that. The argument for shrinking none is that we already put other labels in the margins - cartesian axis  titles, ticks, & tick labels. People making multi-subplot plots are used to leaving space for this between their subplots, and in fact if you make a `layout.grid` (see https://codepen.io/alexcjohnson/pen/NBZYpm/) there's a gap by default between neighboring subplots, which is easy to increase (`layout.grid.ygap`).

I guess *most* of our subplots - whether single-trace like pie or multi-trace like cartesian - stay within their domains, which would argue for shrinking all. The only exceptions I think are cartesian and ternary subplots.

File: src/traces/pie/plot.js [hunk 4]
```diff
@@ -454,6 +501,72 @@ function transformOutsideText(textBB, pt) {
...
+    return false;
+}
+
+function getTitleSpace(trace, plotSize) {
+    var chartHeight = plotSize.h * (trace.domain.y[1] - trace.domain.y[0]);
+    // leave 3/2 * titlefont.size free space. We need at least titlefont.size
+    // space, and the 1/2 * titlefont.size is a small buffer to avoid the text
+    // touching the pie.
```
Comment (alexcjohnson) [inline src/traces/pie/plot.js]: This should be based on `titleBB.height` - since you can include `'multi<br>line<br>titles'`
But a buffer of half the font size is still good.

Actually, in principle anything that depends on the size of text needs to happen in the (potentially async) callback 3rd argument to `convertToTspans`, in order to handle MathJax (`title: '$x+y=z$'` etc). That may throw a bit of a wrench into this whole process, since `scalePies` happens so early on in the process... seems like perhaps all the pie titles need to be rendered upfront, stashed somewhere, and *all* the existing plotting happens in their combined completion callback. Super annoying, I know...

File: test/jasmine/tests/pie_test.js [hunk 5]
```diff
@@ -182,6 +182,53 @@ describe('Pie traces:', function() {
...
+        Plotly.newPlot(gd, [{
+            values: [2, 2, 2, 2],
+            title: 'Pie',
+            titleposition: 'inhole',
+            hole: 0,
+            type: 'pie',
+        }], {height: 300, width: 300})
+        .then(function() {
+            var title = d3.selectAll('.titletext text');
+            expect(title.size()).toBe(0);
```
Comment (alexcjohnson) [inline test/jasmine/tests/pie_test.js]: Mentioned above, but I feel like in this case it'd be better to just push the title outside.

File: test/image/mocks/pie_title_pull.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
+{
+  "data": [
+    {
+      "values": [
+        50, 49, 48, 47, 46, 45, 44, 43, 42, 41
+      ],
+      "pull": [
+        0, 0.5, 0, 0.5, 0, 0.5, 0, 0.5, 0, 0.5
```
Comment (alexcjohnson) [inline test/image/mocks/pie_title_pull.json]: Thanks for considering `pull` in all this - nice touch!

File: test/image/mocks/pie_title_inhole.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
+  "data": [
+    {
+      "values": [955, 405, 360, 310, 295],
+      "labels": ["Mandarin", "Spanish", "English", "Hindi", "Arabic"],
+      "textinfo": "label+percent",
+      "hole": 0.3,
+      "title": "Num. speakers",
```
Comment (alexcjohnson) [inline test/image/mocks/pie_title_inhole.json]: Looks like this one is too small already to trigger scaling down - would you mind making this a series including both large and small titles, and perhaps multiline labels in both cases? (Also multiline for outside labels)

**Comment by alexcjohnson** [comment]
@codrut3 thanks for the PR, this is looking really strong!

One other question that occurred to me is how will these titles interact with outside slice labels? These labels already have known open issues (eg https://github.com/plotly/plotly.js/issues/75) so I won't be surprised if the answer is "not well" and I don't think we need to find a solution to that now, but it's worth keeping in mind and asking whether there's anything simple we could do to improve it in the short term.

File: src/traces/pie/attributes.js (hunk 1 above)
Comment (etpinard) [reply src/traces/pie/attributes.js]: > What you have now corresponds to 'top center' and 'middle center', and it's fine to start with just those two options, but I can certainly see wanting 'top left', 'top right', and 'bottom (left|center|right)' in the future.

> Also I'd remove the dflt here and make it dependent on hole: if it's zero, default to 'top center', otherwise 'middle center'.

voting :+1:

File: src/traces/pie/plot.js (hunk 3 above)
Comment (etpinard) [reply src/traces/pie/plot.js]: > The only exceptions I think are cartesian and ternary subplots.

Polar axes are also drawn out-of-their domains (for most cases that only implies angularaxis tick and tick labels). In other words only the _plotting_ area is constraint by the domain.

If I understand correctly, long outside labels (e.g. https://rreusser.github.io/plotly-mock-viewer/#pie_style_arrays) do not have an effect on the margins, but arguably _could_ (just like cartesian axis tick labels now have an `automargin:true` option). 

All in all, I'm leaning on the side on consistency with outside labels, that is don't touch the margins but shrink the pies.



> I don't think we want to treat single-pie-per-column differently from multiple

I agreed :100: % here. So yeah, let's shrink them all!

**Comment by codrut3** [comment]
Hi @alexcjohnson @etpinard ,

 Thank you for your comments and for taking the time to review my PR! I will probably need a few days to make the changes. I will add all the options that you mentioned to `titleposition`, pre-render the title and shrink the pie to position it. But I need to think about MathJax. Can't I just set `data-notex`? :smile:
 
Also, the title may intersect outside labels. I also noticed that outside labels can go outside the plot or overlap other pies. This is, like, a whole issue on its own. I will think about this too.

**Comment by alexcjohnson** [comment]
> Can't I just set `data-notex`?

Actually that's a good point, since we don't support LaTeX in pie slices either for now (see the [mathjax test image](https://github.com/plotly/plotly.js/blob/master/test/image/baselines/mathjax.png)), I wouldn't be opposed to `data-notex` here. I suspect there's not *too* much intersection between pie users and LaTeX users... :neckbeard:

**Comment by codrut3** [comment]
Hi,
I've made some changes. Can you have a look and tell me if it is fine so far?

I added the following options to `titleposition`: top left, top center, top right, middle center, bottom left, bottom center, bottom right. I did not add middle left and right, because I am not sure how they would look.

Also, I added unit tests that:
- check correct placement of multiline titles in all 7 cases,
- check correct placement when some slices are pulled,
- check correct scaling of title when placed in the hole ('middle center'),
- check correct placing of very large title.

If this is fine, next I will try to do something about titles intersecting outside labels.

Also, note that placing the title to the left (or right) means placing it on the left (or right) margin of the pie (`x = center.x - radius`). So not the left edge of the subplot, which would be further away. Is this good? What do you think?

File: src/traces/pie/attributes.js (hunk 1 above)
Comment (codrut3) [reply src/traces/pie/attributes.js]: Done, added all options mentioned above.

File: src/traces/pie/plot.js (hunk 3 above)
Comment (codrut3) [reply src/traces/pie/plot.js]: Done. I always shrink the pies to gain space for the title.

File: src/traces/pie/plot.js (hunk 4 above)
Comment (codrut3) [reply src/traces/pie/plot.js]: I handled multiline titles.
Making it work for MathJax is much harder. Maybe I could just use `data-notex`?

File: test/jasmine/tests/pie_test.js (hunk 5 above)
Comment (codrut3) [reply test/jasmine/tests/pie_test.js]: Done.

File: test/image/mocks/pie_title_pull.json (hunk 6 above)
Comment (codrut3) [reply test/image/mocks/pie_title_pull.json]: Thanks!

File: test/image/mocks/pie_title_inhole.json (hunk 7 above)
Comment (codrut3) [reply test/image/mocks/pie_title_inhole.json]: Done. I added other mocks using multiline titles and scaling.

File: test/jasmine/tests/pie_test.js [hunk 8]
```diff
@@ -183,52 +183,236 @@ describe('Pie traces:', function() {
...
+            title: 'Test<BR>Title',
+            titleposition: 'middle center',
+            titlefont: {
+                size: 12
+            },
+            hole: 0,
+            type: 'pie',
+            textinfo: 'none'
+        }], {height: 300, width: 300})
+        .then(_verifyTitle(false, false, true, false, true))
```
Comment (alexcjohnson) [inline test/jasmine/tests/pie_test.js]: `_verifyTitle` is beautifully DRY 🌴 ! But you're running it synchronously here, because the argument to `then` is the *result* of `_verifyTitle`, not a function to call when the `newPlot` promise resolves. Which turns out to be ok, as this plot (with no MathJax or maps) renders synchronously. My preferred way to handle this is to return a function from `_verifyTitle`:

```js
function _verifyTitle(checkLeft, checkRight, checkTop, checkBottom, checkMiddleX) {
    return function() {
        var title = d3.selectAll('.titletext text');
        ...
    };
}
```

File: src/traces/pie/defaults.js [hunk 9]
```diff
@@ -65,19 +65,17 @@ module.exports = function supplyDefaults(traceIn, traceOut, defaultColor, layout
         }
     }
 
+    handleDomainDefaults(traceOut, layout, coerce);
+
+    var hole = coerce('hole');
     var title = coerce('title');
     if(title) {
         var titlePosition = coerce('titleposition');
```
Comment (alexcjohnson) [inline src/traces/pie/defaults.js]: I still think we should make the default be `'middle center'` when there's a `hole` (ie `var titlePosition = coerce('titleposition', hole ? 'middle center' : 'top center');` and remove `attributes.titleposition.dflt`) - or maybe when `hole > 0.5` or something... would be strange to put the title in the center when the hole is very small!

But I'm not wedded to that idea; if you've considered it and still think `'top center'` should be the default in all cases we can keep it that way.

File: src/traces/pie/plot.js (hunk 4 above)
Comment (alexcjohnson) [reply src/traces/pie/plot.js]: per https://github.com/plotly/plotly.js/pull/2987#issuecomment-421176694 - yes, just use `data-notex`.

**Comment by alexcjohnson** [comment]
@codrut3 thanks for the changes, they look fantastic! Lovely tests and very thorough test images 🏆 

> Also, note that placing the title to the left (or right) means placing it on the left (or right) margin of the pie (`x = center.x - radius`). So not the left edge of the subplot, which would be further away. Is this good? What do you think?

Agreed 💯 - thanks for pointing this out.

> If this is fine, next I will try to do something about titles intersecting outside labels.

You're certainly welcome to do that here, but that can also be a separate PR (ideally addressing #75 too).

So the only things I'd still like to see before merging are `_verifyTitle` https://github.com/plotly/plotly.js/pull/2987#discussion_r219370769, `data-notex` https://github.com/plotly/plotly.js/pull/2987#discussion_r219372847 and finalizing the default `titleposition` https://github.com/plotly/plotly.js/pull/2987#discussion_r219372162

**Comment by codrut3** [comment]
Hi @alexcjohnson ,
Thank you for the review! I've changed `_verifyTitle` and `titleposition` as you say. However, what should I do about `data-notex`? I set it for the title on line 323. Should I write a test?

**Comment by etpinard** [comment]
Amazing PR @codrut3 

Adding this to the `1.42.0` milestone.

**Comment by alexcjohnson** [comment]
Ah sorry, I hadn't noticed that `data-notex` was already in there, I thought it was still a "should I do this" question. Nothing further is needed. Fantastic work @codrut3, thanks for the quick updates! 💃 

@etpinard whenever you're ready to start merging 1.42 features go for it!

**Comment by etpinard** [comment]
> whenever you're ready to start merging 1.42 features go for it!

We'll make one more release under 1.41.x next week, then start prepping 1.42.0.

**Comment by nicolaskruchten** [comment]
Wow, this is really exciting! Hey @ebellumat once this is done I actually would consider adding first-class multi-pie support to PivotTable.js and react-pivottable :)

**Comment by etpinard** [comment]
Starting merging things for 1.42.0 now.

**Comment by ebellumat** [comment]
> Wow, this is really exciting! Hey @ebellumat once this is done I actually would consider adding first-class multi-pie support to PivotTable.js and react-pivottable :)

Wow, I was waiting for this, I did not imagine it would be so fast! Congratulations to those involved. And Thanks to @nicolaskruchten to remember that pull request in PivotTable.js
